test-data/unit/check-classes.test

```diff
@@ -894,6 +892,281 @@ a.f = a.f # E: Property "f" defined in "A" is read-only
 a.f.x # E: "int" has no attribute "x"
 [builtins fixtures/property.pyi]
 
+-- Descriptors
+-- -----------
+
+
+[case testAccessingNonDataDescriptor]
+class D():
+    def __get__(self, instance: Any, owner: Any) -> str:
+        return 'some string'
+
+class A:
+    f = D()
+
+a = A() # type: A
+reveal_type(a.f)  # E: Revealed type is 'builtins.str'
+
+[case testSettingDataDescriptor]
+class D():
+    def __get__(self, instance: Any, owner: Any) -> str:
+        return 'some string'
+
+    def __set__(self, instance: Any, value: str) -> None:
+        pass
+
+class A:
+    f = D()
+
+a = A() # type: A
+a.f = ''
+a.f = 1 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+[case testReadingDescriptorWithoutDunderGet]
+from typing import Union
+
+class D():
+    def __set__(self, instance: Any, value: str) -> None:
+        pass
+
+class A:
+    f = D()  # type: Union[D, str]
```

Comment:
This test feels fishy. Without this `# type` the assignment `self.f = 'some string'` should pass, without a `# type: str` there. And the revealed type for a.f should be D. It feels as if the explicit annotations are trying to paper over something missing in the implementation.